File: cpp/tradutor.cpp
#include <iostream>
#include <fstream>
#include <stack>
#include <string>
#include <stdexcept>

int obterPrecedencia(char op) {
    if (op == '^')
        return 3;
    if (op == '*' || op == '/')
        return 2;
    if (op == '+' || op == '-')
        return 1;
    return 0;
}

std::string infixaParaPosfixa(const std::string& infixa) {
    std::stack<char> pilha;
    std::string posfixa = "";

    for (char token : infixa) {
        if (isalnum(token)) {
            posfixa += token;
        } else if (token == '(') {
            pilha.push(token);
        } else if (token == ')') {
            while (!pilha.empty() && pilha.top() != '(') {
                posfixa += pilha.top();
                pilha.pop();
            }
            if (!pilha.empty() && pilha.top() == '(') {
                pilha.pop();
            } else {
                throw std::runtime_error("Erro: Parênteses desbalanceados na expressão.");
            }
        } else {
            while (!pilha.empty() && obterPrecedencia(pilha.top()) >= obterPrecedencia(token)) {
                posfixa += pilha.top();
                pilha.pop();
            }
            pilha.push(token);
        }
    }

    while (!pilha.empty()) {
        if (pilha.top() == '(') {
             throw std::runtime_error("Erro: Parênteses desbalanceados na expressão.");
        }
        posfixa += pilha.top();
        pilha.pop();
    }

    return posfixa;
}

int main(int argc, char* argv[]) {
    std::string arquivoEntradaNome;
    std::string arquivoSaidaNome;

    if (argc == 3) {
        arquivoEntradaNome = argv[1];
        arquivoSaidaNome = argv[2];
    } else if (argc == 1) {
        std::cout << "Digite o nome do arquivo de entrada: ";
        std::cin >> arquivoEntradaNome;
        std::cout << "Digite o nome do arquivo de saida: ";
        std::cin >> arquivoSaidaNome;
    } else {
        std::cerr << "Uso: tradutor_cpp <arquivo_entrada> <arquivo_saida>" << std::endl;
        std::cerr << "Ou execute sem argumentos para modo interativo." << std::endl;
        return 1;
    }

    std::ifstream arquivoEntrada(arquivoEntradaNome);
    std::ofstream arquivoSaida(arquivoSaidaNome);

    if (!arquivoEntrada.is_open()) {
        std::cerr << "Erro ao abrir o arquivo de entrada: " << arquivoEntradaNome << std::endl;
        return 1;
    }

    if (!arquivoSaida.is_open()) {
        std::cerr << "Erro ao criar o arquivo de saida: " << arquivoSaidaNome << std::endl;
        return 1;
    }

    std::string linha;
    while (std::getline(arquivoEntrada, linha)) {
        if (linha.empty()) continue;
        try {
            std::string posfixa = infixaParaPosfixa(linha);
            arquivoSaida << posfixa << std::endl;
        } catch (const std::runtime_error& e) {
            std::cerr << "Erro ao processar a expressao '" << linha << "': " << e.what() << std::endl;
        }
    }

    std::cout << "Tradução concluida. Resultados em: " << arquivoSaidaNome << std::endl;

    arquivoEntrada.close();
    arquivoSaida.close();

    return 0;
}

File: java/Tradutor.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Tradutor {

    private static int obterPrecedencia(char op) {
        switch (op) {
            case '^':
                return 3;
            case '*':
            case '/':
                return 2;
            case '+':
            case '-':
                return 1;
        }
        return 0;
    }

    public static String infixaParaPosfixa(String infixa) throws IllegalArgumentException {
        StringBuilder posfixa = new StringBuilder();
        Stack<Character> pilha = new Stack<>();

        for (char token : infixa.toCharArray()) {
            if (Character.isLetterOrDigit(token)) {
                posfixa.append(token);
            } else if (token == '(') {
                pilha.push(token);
            } else if (token == ')') {
                while (!pilha.isEmpty() && pilha.peek() != '(') {
                    posfixa.append(pilha.pop());
                }
                if (!pilha.isEmpty() && pilha.peek() == '(') {
                    pilha.pop();
                } else {
                    throw new IllegalArgumentException("Erro: Parênteses desbalanceados na expressão: " + infixa);
                }
            } else {
                while (!pilha.isEmpty() && obterPrecedencia(pilha.peek()) >= obterPrecedencia(token)) {
                    posfixa.append(pilha.pop());
                }
                pilha.push(token);
            }
        }

        while (!pilha.isEmpty()) {
            if (pilha.peek() == '(') {
                 throw new IllegalArgumentException("Erro: Parênteses desbalanceados na expressão: " + infixa);
            }
            posfixa.append(pilha.pop());
        }

        return posfixa.toString();
    }

    public static void main(String[] args) {
        String arquivoEntradaNome;
        String arquivoSaidaNome;
        BufferedReader consoleReader = new BufferedReader(new InputStreamReader(System.in));

        try {
            if (args.length == 2) {
                arquivoEntradaNome = args[0];
                arquivoSaidaNome = args[1];
            } else if (args.length == 0) {
                System.out.print("Digite o nome do arquivo de entrada: ");
                arquivoEntradaNome = consoleReader.readLine();
                System.out.print("Digite o nome do arquivo de saida: ");
                arquivoSaidaNome = consoleReader.readLine();
            } else {
                System.err.println("Uso: java Tradutor <arquivo_entrada> <arquivo_saida>");
                System.err.println("Ou execute sem argumentos para modo interativo.");
                return;
            }

            try (BufferedReader leitor = new BufferedReader(new FileReader(arquivoEntradaNome));
                 BufferedWriter escritor = new BufferedWriter(new FileWriter(arquivoSaidaNome))) {

                String linha;
                while ((linha = leitor.readLine()) != null) {
                    if (linha.trim().isEmpty()) continue;
                    try {
                        String posfixa = infixaParaPosfixa(linha.trim());
                        escritor.write(posfixa);
                        escritor.newLine();
                    } catch (IllegalArgumentException e) {
                        System.err.println("Erro ao processar a expressao '" + linha + "': " + e.getMessage());
                        escritor.write("Erro: " + e.getMessage() + " na expressao: " + linha);
                        escritor.newLine();
                    }
                }
                System.out.println("Tradução concluida. Resultados em: " + arquivoSaidaNome);

            } catch (IOException e) {
                System.err.println("Erro de E/S: " + e.getMessage());
            }
        } catch (IOException e) {
            System.err.println("Erro ao ler entrada do console: " + e.getMessage());
        }
    }
}
